### ISSUE TITLE
Uploading certain kinds of videos will put zou-events in an unreliable state

### ISSUE THREAD

**Issue by cuibonobo** [cgwire/kitsu#707]
I did a fresh install of Zou and Kitsu yesterday to try it out and so far I really like it. I did want to report a bug that I'm experiencing when uploading certain videos. I've attached an example video to see if someone else can reproduce this behavior.

Here's my `/api/status` output for good measure:
```json
{"name": "Zou", "version": "0.12.76", "database-up": true, "key-value-store-up": true, "event-stream-up": true, "job-queue-up": true}
```

When I drag this video onto a shot, it uploads pretty quickly and I see the loading animation for the preview. However, the preview never updates. After several minutes I refresh the page to see if the preview is done but instead it says, "This preview is broken." In addition, if I check the console there is now an error that says:
```text
Firefox can’t establish a connection to the server at wss://<server_domain>/socket.io/?EIO=3&transport=websocket&sid=a8f07c54399543609c2e250f5d735801. 
```

If I run `journalctl -u zou-events.service`, I see the following:
```text
May 27 15:46:36 kitsu gunicorn[7722]: message handler error
May 27 15:46:36 kitsu gunicorn[7722]: Traceback (most recent call last):
May 27 15:46:36 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/engineio/server.py", line 520, in _trigger_event
May 27 15:46:36 kitsu gunicorn[7722]:     return self.handlers[event](*args)
May 27 15:46:36 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/socketio/server.py", line 697, in _handle_eio_message
May 27 15:46:36 kitsu gunicorn[7722]:     self._handle_connect(sid, pkt.namespace)
May 27 15:46:36 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/socketio/server.py", line 592, in _handle_connect
May 27 15:46:36 kitsu gunicorn[7722]:     self.environ[sid])
May 27 15:46:36 kitsu gunicorn[7722]: KeyError: 'bf6df46435c142df8c09ba896150d961'
May 27 15:48:01 kitsu gunicorn[7722]: Traceback (most recent call last):
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/gevent/pywsgi.py", line 970, in handle_one_response
May 27 15:48:01 kitsu gunicorn[7722]:     self.run_application()
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/geventwebsocket/handler.py", line 75, in run_application
May 27 15:48:01 kitsu gunicorn[7722]:     self.run_websocket()
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/geventwebsocket/handler.py", line 52, in run_websocket
May 27 15:48:01 kitsu gunicorn[7722]:     list(self.application(self.environ, lambda s, h, e=None: []))
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/flask/app.py", line 2463, in __call__
May 27 15:48:01 kitsu gunicorn[7722]:     return self.wsgi_app(environ, start_response)
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/flask_socketio/__init__.py", line 45, in __call__
May 27 15:48:01 kitsu gunicorn[7722]:     return super(_SocketIOMiddleware, self).__call__(environ,
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/engineio/middleware.py", line 60, in __call__
May 27 15:48:01 kitsu gunicorn[7722]:     return self.engineio_app.handle_request(environ, start_response)
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/socketio/server.py", line 526, in handle_request
May 27 15:48:01 kitsu gunicorn[7722]:     return self.eio.handle_request(environ, start_response)
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/engineio/server.py", line 348, in handle_request
May 27 15:48:01 kitsu gunicorn[7722]:     socket = self._get_socket(sid)
May 27 15:48:01 kitsu gunicorn[7722]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/engineio/server.py", line 536, in _get_socket
May 27 15:48:01 kitsu gunicorn[7722]:     raise KeyError('Session is disconnected')
May 27 15:48:01 kitsu gunicorn[7722]: KeyError: 'Session is disconnected'
May 27 15:48:01 kitsu gunicorn[7722]: 2021-05-27T15:48:01Z {'REMOTE_ADDR': '127.0.0.1', 'REMOTE_PORT': '33846', 'HTTP_HOST': '<server_domain>', (hidden keys: 35)} fai>May 27 15:51:39 kitsu systemd[1]: Stopping Gunicorn instance to serve the Zou Events API...
```

Despite this error message and not being able to connect to the websocket, `service zou-events status` still shows that the event service is active and running, which is why I say it's in an unreliable state. The `/api/status` endpoint also shows `"event-stream-up": true` even though I'm not able to connect anymore.

As for what caused the error in the first place, here's the output of `journalctl -u zou-jobs.service`:
```text
May 27 15:50:01 kitsu rq[2623]: 15:50:01 default: zou.app.services.preview_files_service.prepare_and_store_movie('57b5663a-e192-49e5-b368-35656e35fc89', '/tmp/zou/57b5663a-e192-49e5-b368-35656e35fc89.mp4.tmp') (b1a794f3-2a22-45f8-94f0-2808d66dea4b)
May 27 15:50:12 kitsu rq[7759]: [2021-05-27 15:50:12,698] ERROR in preview_files_service: b"ffmpeg version 4.3.2-0+deb11u1ubuntu1 Copyright (c) 2000-2021 the FFmpeg developers\n  built with gcc 10 (Ubuntu 10.2.1-20ubuntu1)\n  configuration: --prefix=/usr --extra-version=0+deb11u1ubuntu1 --toolchain=hardened --libdir=/usr/li>
May 27 15:50:12 kitsu rq[7759]: [2021-05-27 15:50:12,709] ERROR in preview_files_service: failed
May 27 15:50:12 kitsu rq[7759]: Traceback (most recent call last):
May 27 15:50:12 kitsu rq[7759]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/zou/app/services/preview_files_service.py", line 98, in prepare_and_store_movie
May 27 15:50:12 kitsu rq[7759]:     ) = movie.normalize_movie(
May 27 15:50:12 kitsu rq[7759]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/zou/utils/movie.py", line 109, in normalize_movie
May 27 15:50:12 kitsu rq[7759]:     stream.run(quiet=False, capture_stderr=True, overwrite_output=True)
May 27 15:50:12 kitsu rq[7759]:   File "/opt/zou/zouenv/lib/python3.9/site-packages/ffmpeg/_run.py", line 325, in run
May 27 15:50:12 kitsu rq[7759]:     raise Error('ffmpeg', out, err)
May 27 15:50:12 kitsu rq[7759]: ffmpeg._run.Error: ffmpeg error (see stderr output for detail)
May 27 15:50:12 kitsu rq[7759]: 15:50:12 default: Job OK (b1a794f3-2a22-45f8-94f0-2808d66dea4b)
May 27 15:50:12 kitsu rq[7759]: 15:50:12 Result is kept for 500 seconds
```

The output of the `ffmpeg` job (truncated above) is this:
```text
ffmpeg version 4.3.2-0+deb11u1ubuntu1 Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 10 (Ubuntu 10.2.1-20ubuntu1)
  configuration: --prefix=/usr --extra-version=0+deb11u1ubuntu1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample --disable-filter=resample --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/tmp/zou/20ad64c0-bd6b-42fa-8d9a-ade089053ce9.mp4.tmp':
  Metadata:
    major_brand     : isom
    minor_version   : 1
    compatible_brands: isomavc1mp42
    creation_time   : 2021-05-19T17:29:10.000000Z
  Duration: 00:00:11.97, start: 0.000000, bitrate: 1021 kb/s
    Stream #0:0(und): Video: h264 (Constrained Baseline) (avc1 / 0x31637661), yuv420p(tv, smpte170m/smpte170m/bt709), 960x540 [SAR 1:1 DAR 16:9], 922 kb/s, 25 fps, 25 tbr, 25 tbn, 50 tbc (default)
    Metadata:
      creation_time   : 2021-05-19T17:29:04.000000Z
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 102 kb/s (default)
    Metadata:
      creation_time   : 2021-05-19T17:29:04.000000Z
Please use -b:a or -b:v, -b is ambiguous
Stream mapping:
  Stream #0:0 -> #0:0 (h264 (native) -> h264 (libx264))
  Stream #0:1 -> #0:1 (aac (native) -> aac (native))
Press [q] to stop, [?] for help
[libx264 @ 0x555e00c24700] using SAR=1/1
[libx264 @ 0x555e00c24700] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
[libx264 @ 0x555e00c24700] profile High, level 5.0, 4:2:0, 8-bit
[libx264 @ 0x555e00c24700] 264 - core 160 r3011 cde9a93 - H.264/MPEG-4 AVC codec - Copyleft 2003-2020 - http://www.videolan.org/x264.html - options: cabac=1 ref=5 deblock=1:0:0 analyse=0x3:0x113 me=hex subme=8 psy=1 psy_rd=1.00:0.00 mixed_ref=1 me_range=16 chroma_me=1 trellis=2 8x8dct=1 cqm=0 deadzone=21,11 fast_pskip=1 chroma_qp_offset=-2 threads=1 lookahead_threads=1 sliced_threads=0 nr=0 decimate=1 interlaced=0 bluray_compat=0 constrained_intra=0 bframes=3 b_pyramid=2 b_adapt=1 b_bias=0 direct=3 weightb=1 open_gop=0 weightp=2 keyint=250 keyint_min=24 scenecut=40 intra_refresh=0 rc_lookahead=50 rc=abr mbtree=1 bitrate=28000 ratetol=1.0 qcomp=0.60 qpmin=0 qpmax=69 qpstep=4 ip_ratio=1.40 aq=1:1.00
Output #0, mp4, to '/tmp/zou/20ad64c0-bd6b-42fa-8d9a-ade089053ce9.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 1
    compatible_brands: isomavc1mp42
    encoder         : Lavf58.45.100
    Stream #0:0(und): Video: h264 (libx264) (avc1 / 0x31637661), yuv420p(bt709, progressive), 1920x1080 [SAR 1:1 DAR 16:9], q=-1--1, 28000 kb/s, 24 fps, 12288 tbn, 24 tbc (default)
    Metadata:
      creation_time   : 2021-05-19T17:29:04.000000Z
      encoder         : Lavc58.91.100 libx264
    Side data:
      cpb: bitrate max/min/avg: 0/0/28000000 buffer size: 0 vbv_delay: N/A
    Stream #0:1(eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 128 kb/s (default)
    Metadata:
      creation_time   : 2021-05-19T17:29:04.000000Z
      encoder         : Lavc58.91.100 aac
frame=   49 fps= 31 q=0.0 size=       0kB time=00:00:01.45 bitrate=   0.3kbits/s speed=0.927x
```

I can see in `/tmp/zou` that the file is uploaded to a tmp location, but then `ffmpeg` is unable to create a valid output (the error files always have a size of `48`).
![image](https://user-images.githubusercontent.com/391241/119862225-48ea4400-bee6-11eb-98ad-a57031543de2.png)

For what it's worth, I'm not 100% sure if the bad upload causes me to disconnect from the event service, but I've seen these 2 problems together enough times to think they are perhaps related.

https://user-images.githubusercontent.com/391241/119859657-95805000-bee3-11eb-820e-187d68dd8768.mp4

**Comment by cuibonobo** [comment]
Well, now I'm not able to establish a reliable websocket connection at all, so the issue with this video is not related to not being able to establish a reliable websocket connection.

**Comment by frankrousseau** [comment]
Hello @cuibonobo,
Thank you for your interest in Kitsu. I don't know what to say, the errors are not explicit. For the websocket part, you should check your Nginx installation. For the video, it may be due to folder permissions.

**Comment by cuibonobo** [comment]
I don't thing the video issue is permissions. I'm able to upload other kinds of videos (with different codecs and such) and those will preview fine. The ffmpeg process seems to be halting but I'm not sure why. I'll try to assemble the ffmpeg command from the code and run it manually on the server to see if I get the same result.

In regards to the websocket problem, here's an nginx config snippet:
```nginx
    location /socket.io {
        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_pass http://localhost:5001;
    }
```

Does something need to be added to the configuration to support TLS connections?

**Comment by cuibonobo** [comment]
I made a little script based on [this bit of code](https://github.com/cgwire/zou/blob/9ed714b47e71b56c2fafab46dbc5cb68d0fc8516/zou/utils/movie.py#L92-L108) and then used `ffmpeg.compile(stream)` to get the the ffmpeg command being run on the command line:
```bash
ffmpeg -i /tmp/zou/33d9712d-d093-478f-89e6-537bea667535.mp4.tmp -map 0:v -map 0:a -f mp4 -b:v 28M -color_primaries 1 -color_trc 1 -colorspace 1 -movflags +faststart -pix_fmt yuv420p -preset slow -r 24 -s 1920x1080 -vcodec libx264 /tmp/zou/myoutput.mp4
```

When I run this I get the same stderr output as my original comment, but now I can see that the exit code is `137`. This seems to suggest that [the ffmpeg process is being killed](https://stackoverflow.com/questions/19548027/ffmpeg-closes-with-return-code-137) by the operating system because of CPU or memory limits. To test this, I raised the memory on my server from 1GB to 2GB and that fixed the video issue.

I was hoping that raising the memory limit would also fix the websocket problem, but no dice.

**Comment by cuibonobo** [comment]
I plugged `wss://<my_domain>/socket.io/?transport=websocket` into [this websockets echo test](https://www.websocket.org/echo.html) and it seems like my actual server is working fine.

After doing a bit more research it seems like the websocket error I'm seeing in my console is due to a websocket connection being closed before it has a chance to be fully established. I made an example JSBin to show the basic concept: https://jsbin.com/tumuquy/edit?html,console. Also, it seems like this error is coming from `vue-websocket` so I don't think it's anything that Kitsu is doing.

I'm going to go ahead and close this issue. Thanks for letting me rubber duck my troubleshooting steps on here. Hopefully someone else finds it helpful.
